torchsde/_core/methods/milstein.py

```diff
@@ -12,30 +12,74 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS
+import abc
+import math
+
+import torch
+
+from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS, METHOD_OPTIONS
 
 from .. import base_solver
 
 
-class Milstein(base_solver.BaseSDESolver):
+class BaseMilstein(base_solver.BaseSDESolver, metaclass=abc.ABCMeta):
     strong_order = 1.0
     weak_order = 1.0
-    sde_type = SDE_TYPES.ito
     noise_types = (NOISE_TYPES.additive, NOISE_TYPES.diagonal, NOISE_TYPES.scalar)
     levy_area_approximations = LEVY_AREA_APPROXIMATIONS.all()
 
+    def __init__(self, options, **kwargs):
+        if METHOD_OPTIONS.grad_free not in options:
+            options[METHOD_OPTIONS.grad_free] = False
+        super(BaseMilstein, self).__init__(options=options, **kwargs)
+
+    @abc.abstractmethod
+    def v_term(self, I_k, dt):
+        raise NotImplementedError
+
     def step(self, t0, y0, dt):
         assert dt > 0, 'Underflow in dt {}'.format(dt)
 
-        I_k = self.bm(t0, t0 + dt)
-        v = [delta_bm_ ** 2. - dt for delta_bm_ in I_k]
+        t1 = t0 + dt
+
+        I_k = self.bm(t0, t1)
+        v = self.v_term(I_k, dt)
 
         f_eval = self.sde.f(t0, y0)
-        g_prod_eval = self.sde.g_prod(t0, y0, I_k)
-        gdg_prod_eval = self.sde.gdg_prod(t0, y0, v)
+        g_eval = self.sde.g(t0, y0)
+        g_prod_eval = self.sde.prod(g_eval, I_k)
+
+        if self.options[METHOD_OPTIONS.grad_free]:
+            g_prod_eval_v = self.sde.prod(g_eval, v)
+            sqrt_dt = torch.sqrt(dt) if isinstance(dt, torch.Tensor) else math.sqrt(dt)
+            y0_prime = [
+                y0_ + dt * f_eval_ + g_eval_ * sqrt_dt
+                for y0_, f_eval_, g_eval_ in zip(y0, f_eval, g_eval)
+            ]            
+            g_prod_eval_prime = self.sde.g_prod(t1, y0_prime, v)
```

Comment:
And according to DiffEq here's `t0` being used (But with those fixes slope is still 0.6).